File: Medium/python/1249. Minimum Remove to Make Valid Parentheses.py
"""
Given a string s of '(' , ')' and lowercase English characters.

Your task is to remove the minimum number of parentheses ( '(' or ')', in any positions ) so that the resulting parentheses string is valid and return any valid string.

Formally, a parentheses string is valid if and only if:

It is the empty string, contains only lowercase characters, or
It can be written as AB (A concatenated with B), where A and B are valid strings, or
It can be written as (A), where A is a valid string.
"""

class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        ans = ""
        stack = []

        for ind, char in enumerate(s):
            if char == ")" and stack:
                stack.pop()
                ans += char
            elif char =="(":
                stack.append(")")
                ans += char
            elif not char == ")":
                ans += char

        while stack:
            ind = ans.rindex("(")
            ans = ans[:ind] + ans[ind+1:]
            stack.pop()

        return ans

print(Solution().minRemoveToMakeValid("lee(t(c)o)de)"))

File: Medium/Javascript/1249. Minimum Remove to Make Valid Parentheses.js
/**Given a string s of '(' , ')' and lowercase English characters.

Your task is to remove the minimum number of parentheses ( '(' or ')', in any positions ) so that the resulting parentheses string is valid and return any valid string.

Formally, a parentheses string is valid if and only if:

It is the empty string, contains only lowercase characters, or
It can be written as AB (A concatenated with B), where A and B are valid strings, or
It can be written as (A), where A is a valid string. */

/**
 * @param {string} s
 * @return {string}
 */
var minRemoveToMakeValid = function (s) {
  let ans = "";
  let stack = [];

  for (let char of s) {
    if (char === ")" && stack.length >= 1) {
      ans += char;
      stack.pop();
    } else if (char === "(") {
      ans += char;
      stack.push(")");
    } else if (char != ")") {
      ans += char;
    }
  }

  while (stack.length >= 1) {
    let ind = ans.lastIndexOf("(");
    ans = ans.slice(0, ind) + ans.slice(ind + 1, ans.length + 1);
    stack.pop()
  }

  return ans
};


console.log(minRemoveToMakeValid("))(("))